Bug fix when `nstates=1`; when `expects is None`; small changes to plotting

cyrus-baker opened the pull request:
1. fix bug when plot_expects when expects is None; 
2. fix dim calculation in PropagatorInfidelity; 
3. fix bug when initial state is only one in CoherentInfidelity

Commit:
fix(CoherentInfidelity,-PropagatorInfidelity,-plot_expects): 1. fix bug when plot expects in propagator where expects is None; 2. fix dim calculation in PropagatorInfidelity; 3. fix bug when initial state is only one in CoherentInfidelity

dkweiss31 commented on qontrol/cost.py:
```diff
@@ -375,16 +375,16 @@ def __call__(
         overlaps = self.target_states.dag() @ result.final_state
         if not isket(result.final_state):
-            overlaps = overlaps.trace()
+            overlaps = overlaps.trace(axis1=-1, axis2=-2)
```
I'm not sure this is necessary, see [here](https://github.com/dynamiqs/dynamiqs/blob/main/dynamiqs/qarrays/dense_qarray.py#L87-L88) for the function this calls which already specifies the last two axes

cyrus-baker replied:
Oh yes, this bug is caused by `squeeze`, not by `trace`

dkweiss31 commented on qontrol/cost.py:
```diff
@@ -375,16 +375,16 @@ def __call__(
             # the trace is always real and positive. Included here only for
             # completeness.
-            overlaps_avg = jnp.mean(jnp.squeeze(overlaps), axis=-1)
+            overlaps_avg = jnp.mean(jnp.squeeze(overlaps, axis=[-1, -2]), axis=-1)
```
Good catch!

dkweiss31 commented on qontrol/cost.py:
```diff
@@ -401,7 +401,7 @@ def __call__(
         parameters: dict | Array,  # noqa ARG002
     ) -> tuple[tuple[Array, Array]]:
-        dim = jnp.prod(*result.final_propagator.dims)
+        dim = jnp.prod(jnp.array(result.final_propagator.dims))
```
Would the simpler
```suggestion
        dim = jnp.prod(result.final_propagator.dims)
```
work?

cyrus-baker replied:
No, `result.final_propagator.dims` is a tuple. However, jax require only `jnp.ndarray` and do not accept implict convertion

dkweiss31 commented on qontrol/plot.py:
```diff
@@ -100,9 +100,10 @@ def plot_expects(
-    expect_idxs = np.ndindex(*expects.shape[:-1])
-    for expect_idx in expect_idxs:
-        ax.plot(tsave, np.real(expects[tuple(expect_idx)]))
+    if expects is not None:
+        expect_idxs = np.ndindex(*expects.shape[:-1])
+        for expect_idx in expect_idxs:
+            ax.plot(tsave, np.real(expects[tuple(expect_idx)]))
```
This isn't going to be the ideal solution, since it will leave a blank panel (regardless, good catch!). Maybe instead we could use different default plotters for mesolve, sesolve versus mepropogator, sepropogator?

cyrus-baker replied:
You are right, this is just a temporary solution. I will see if there is a better solution. Maybe instead of using propagators for "expectations", it is better to use "states" and calculate their evolution. But this may lead to changes in semantics and affect the API.

dkweiss31 reviewed (changes requested):
Thanks for this PR @littlebaker! I've left some comments. I'm wondering if you might be able to adjust your solution when `expects is None` to avoid just having an empty plot. Could we use different default plotters for the propogator functionality? Additionally, you've uncovered a few bugs, and I'm wondering if you can add test coverage for those use cases so that they shouldn't show up again?

littlebaker commented:
Sure, I will remove the empty plot when `expects is None`. As for the plotters functionality, I think we may need further discussion.

littlebaker commented:
Sure, I will add some test code

Commit:
test: add test for single target of coherent infidelity and gate plot

Commit:
fix: remove redundant empty plot when exp_ops is None or []

dkweiss31 commented on qontrol/optimize.py:
```diff
@@ -104,6 +104,18 @@ def optimize(
+            plotter = DefaultPlotter()
+        else:
+            plotter = Plotter([plot_fft, plot_controls])
+
```
I like what you're going for here. I think what this is pointing to is that we need one more layer of abstraction. The base `Model` class shouldn't have `exp_ops` as an attribute, since the propagator models don't have it. I'm thinking of something like
```python
class Model(eqx.Module):
    H_function: callable
    tsave_function: callable

    def __call__(
            self,
            parameters: Array | dict,
            method: Method = Tsit5(),  # noqa B008
            gradient: Gradient | None = None,
            options: dq.Options = dq.Options(),  # noqa B008
    ) -> tuple[Result, TimeQArray]:
        raise NotImplementedError


class SolveModel(Model):
    exp_ops: Array | None


class PropagatorModel(Model):
    pass
```

Would you be able to make these changes to the model file? Then the check above could become much simpler.

cyrus-baker replied:
Sure, that will be a good idea.

littlebaker commented:
Ok, sure. I will remove them.

littlebaker commented:
Sorry, wrong operation, I did not finished.

littlebaker commented:
```python
def plot_controls(
    ax: Axes, _expects: Array | None, model: Model, parameters: Array | dict
) -> Axes:
    """Plot the Hamiltonian prefactors, usually corresponding to controls."""
    ax.set_facecolor('none')
    H = model.H_function(parameters)
    tsave = model.tsave_function(parameters)
    controls = get_controls(H, tsave)
    H_labels = [f'$H_{idx}$' for idx in range(len(controls))]
    for idx, control in enumerate(controls):
        ax.plot(tsave, np.real(control) / 2 / np.pi, label=H_labels[idx])
    ax.legend(loc='lower right', framealpha=0.0)
    ax.set_ylabel('pulse amplitude [GHz]')
    ax.set_xlabel('time [ns]')
    return ax
```
why `np.real(control) / 2 / np.pi` has `/ 2 / np.pi` ? I think 2 pi should be multiplied in Hamiltonians?

Commit:
add debug norm print

dkweiss31 commented:
Right, this is a mostly arbitrary choice that the controls included the 2pis, so that I needed to divide by 2pi when plotting. You are right though that not everyone will do this and it can be unexpected for users, so I'm happy to change this default behavior (though one can of course use their own custom `Plotter`

Commit:
fix: 1. delete norm print 2. fix the order of Hamiltonian in `get_controls`

Commit:
refactor: 1. remove /2pi in `plot_controls`. 2. reorganized import in plot.py and optimize.py

Commit:
refactor: 1. add a layer to distinguish SolveModel and PropagatorModel. 2. remove exp_ops in PropagatorModel. 3. fix plotter creation in `optimize.py`

BREAKING CHANGE: Now `sepropagator_model` and `mepropagator_model` no longer accept `exp_ops`

Commit:
fix: add `block_until_ready` for accurate timing

Commit:
fix: remove redundant `block_until_ready`

Commit:
fix: delete __pycache__

Commit:
style: lint the code, delete json and cache

dkweiss31 committed:
nits

dkweiss31 reviewed (approved):
Thanks @littlebaker ! Everything LGTM. I will let the ci run and then we can merge :). Congrats on your first contribution!!

dkweiss31 committed:
Merge branch 'main' into branch_fix
